be/src/exec/cache/cache_operator.cpp

```diff
@@ -0,0 +1,455 @@
+    }
+    bool is_partial() const { return state == PLBS_MISS || state == PLBS_HIT_PARTIAL || state == PLBS_PARTIAL; }
+    void append_chunk(const vectorized::ChunkPtr& chunk) {
+        if (!is_partial()) {
+            return;
+        }
+        // CRITICAL!!!: we should append last empty chunk to the empty buffer to guarantee that pull_chunk method
+        // would fetch a chunk and reset_lane.
+        if (!chunk->is_empty() || (chunks.empty() && chunk->owner_info().is_last_chunk())) {
+            chunks.push_back(chunk);
+            num_rows += chunk->num_rows();
+            num_bytes += chunk->bytes_usage();
+        }
+
+        if (chunk->owner_info().is_last_chunk()) {
+            state = chunk->owner_info().is_last_chunk() ? PLBS_TOTAL : PLBS_PARTIAL;
+        }
+    }
+
+    bool has_chunks() const { return next_chunk_idx < chunks.size(); }
+
+    void set_passthrough() {
+        if (is_partial()) {
+            state = PLBS_PASSTHROUGH;
+        }
+    }
+
+    bool can_release() {
+        return (state == PLBS_POPULATE || state == PLBS_PASSTHROUGH || state == PLBS_HIT_TOTAL) && !has_chunks();
+    }
+
+    ChunkPtr get_next_chunk() {
+        if (UNLIKELY(state == PLBS_POPULATE || state == PLBS_PASSTHROUGH)) {
+            while (UNLIKELY(next_gc_chunk_idx < next_chunk_idx)) {
+                chunks[next_gc_chunk_idx++].reset();
+            }
+            ++next_gc_chunk_idx;
+        }
+        // CRITICAL(by satanson): The method is invoked when cloning chunk, when query cache enabled, chunk
+        // may be accessed by multi-thread, so we must clone a chunk when pull chunk from cache operator.
+        ChunkPtr chunk = std::move(chunks[next_chunk_idx++]->clone_unique());
+        return chunk;
+    }
+};
+
+CacheOperator::CacheOperator(pipeline::OperatorFactory* factory, int32_t driver_sequence, CacheManagerRawPtr cache_mgr,
+                             const CacheParam& cache_param)
+        : pipeline::Operator(factory, factory->id(), factory->get_name(), factory->plan_node_id(), driver_sequence),
+          _cache_mgr(cache_mgr),
+          _cache_param(cache_param),
+          _lane_arbiter(std::make_shared<LaneArbiter>(_cache_param.num_lanes)) {
+    _per_lane_buffers.reserve(_lane_arbiter->num_lanes());
+    for (auto i = 0; i < _lane_arbiter->num_lanes(); ++i) {
+        _per_lane_buffers.push_back(std::make_unique<PerLaneBuffer>());
+    }
+}
+
+static inline vectorized::Chunks remap_chunks(const vectorized::Chunks& chunks, const SlotRemapping& slot_remapping) {
+    std::vector<vectorized::ChunkPtr> new_chunks;
+    new_chunks.reserve(chunks.size());
+    for (auto i = 0; i < chunks.size(); ++i) {
+        const auto& chunk = chunks[i];
+        if (chunk->is_empty()) {
+            new_chunks.push_back(chunk);
+            continue;
+        }
+
+        auto new_chunk = std::make_shared<vectorized::Chunk>();
+        for (auto [slot_id, new_slot_id] : slot_remapping) {
```

Comment:
Some of our operators depend on the order of columns such as local shuffle. Please confirm this question

Reply:
SlotId remaping always happens at cache operator,  post-cache agg always produces chunks with consistent maps from SlotIds to column index.